tests/test_sw_triangle.py

```diff
@@ -38,7 +38,10 @@ def setup_sw(dirname, euler_poincare):
                          Difference('SWPotentialEnstrophy_from_PotentialVorticity',
                                     'SWPotentialEnstrophy_from_RelativeVorticity'),
                          Difference('SWPotentialEnstrophy_from_PotentialVorticity',
-                                    'SWPotentialEnstrophy_from_AbsoluteVorticity')]
+                                    'SWPotentialEnstrophy_from_AbsoluteVorticity'),
+                         MeridionalComponent('u'),
```

Comment:
Can we check we get the right thing here?

Reply:
What kind of check do you have in mind?

Reply:
This is Williamson test 2 isn't it? So we couldn't we check that the meridional and vertical components are zero and the zonal component has the expected maximum?

Reply:
It will be near-zero if you're using super bendy elements (mesh `degree=42`). But I think a check that it's "small" relative to the degree of the mesh would be good enough.

Reply:
Yes, sorry, by "zero" I meant "appropriately small for the mesh size and degree" :)

Reply:
I agree completely and think these are very sensible suggestions. I wanted to do something like this to do the check: 

```u_max = data.groups["u"]["max"][0]
tolerance = 0.05

u_meridional = data.groups["u_meridional"]
assert u_meridional["max"][0] < tolerance * u_max

u_radial = data.groups["u_radial"]
assert u_radial["max"][0] < tolerance * u_max

u_zonal = data.groups["u_zonal"]
assert u_max * (1 - tolerance) < u_zonal["max"][0] < u_max * (1 + tolerance)
```

BUT for some reason `u_max` is huge, like 4e9 m/s (faster than the speed of light!) when it should be about 38 m/s. Not really sure why this is and it doesn't show up when you plot the field.

Not sure what to do in this case!

Reply:
Something like that would _definitely_ show up in the plotted field. How is `max` obtained?

Reply:
Through this `maxify` function in `diagnostics.py`:
```
@staticmethod
    def max(f):
        fmax = op2.Global(1, np.finfo(float).min, dtype=float)
        op2.par_loop(op2.Kernel("""
void maxify(double *a, double *b) {
    a[0] = a[0] < fabs(b[0]) ? fabs(b[0]) : a[0];
}
""", "maxify"), f.dof_dset.set, fmax(op2.MAX), f.dat(op2.READ))
return fmax.data[0]
```

In the debugger though I did:
```
u_array  = np.array(u0.dat.data[:])
u_array.max()
```
and got out a similarly large value, so it does look like there's a DOF with a very high value?

Reply:
your `u_max` is in the HDiv space - its numerical values are not useful!

Reply:
to check the diagnostic you can just use the value that we know `u_max` takes - it is defined in the setup part of the file

Reply:
Of course, thanks Jemma! It seems a bit messy to have to define the value of `u_max` in the `test_sw_setup` part as well as the `setup_sw` function though?

The question whether there is a way of doing this that isn't messy! Another way could be to open the netcdf file at the end of the setup file and add a one-off `u_max`, something like
```
data = Dataset("results/"+dirname+"/sw/diagnostics.nc", 'a')
data.createGroup('umax')
data['umax'].createDimension('value', 1)
data['umax'].createVariable('umax', float ('value',))
data['umax']['umax'][0] = u_max
```
but this also seems messy!

Reply:
I don't think we should mess with the netCDF file. Can't you just move the lines defining `u_max` outside of the function definitions so that all functions can see them?

Reply:
Yes! Brain not working today...